src/goto-programs/remove_virtual_functions.cpp

```diff
@@ -170,36 +173,64 @@ void remove_virtual_functionst::remove_virtual_function(
   goto_programt new_code_calls;
   goto_programt new_code_gotos;
 
+  // Get a pointer from which we can extract a clsid.
+  // If it's already a pointer to an object of some sort, just use it;
+  // if it's void* then use the parent of all possible candidates,
+  // which by the nature of get_functions happens to be the last candidate.
+
+  exprt this_expr=code.arguments()[0];
+  assert(this_expr.type().id()==ID_pointer &&
+         "Non-pointer this-arg in remove-virtuals?");
+  const auto& points_to=this_expr.type().subtype();
+  if(points_to==empty_typet())
+  {
+    symbol_typet symbol_type(functions.back().class_id);
+    this_expr=typecast_exprt(this_expr, pointer_typet(symbol_type));
+  }
+  exprt deref=dereference_exprt(this_expr, this_expr.type().subtype());
+  exprt c_id2=build_class_identifier(deref);
+
+  goto_programt::targett last_function;
+
   for(functionst::const_iterator
       it=functions.begin();
       it!=functions.end();
       it++)
   {
-    // call function
     goto_programt::targett t1=new_code_calls.add_instruction();
-    t1->make_function_call(code);
-    to_code_function_call(t1->code).function()=it->symbol_expr;
+    if(it->symbol_expr.get_identifier()!=irep_idt())
+    {
+      // call function
+      t1->make_function_call(code);
+      auto& newcall=to_code_function_call(t1->code);
+      newcall.function()=it->symbol_expr;
+      pointer_typet need_type(symbol_typet(it->symbol_expr.get(ID_C_class)));
+      if(newcall.arguments()[0].type()!=need_type)
```

Comment:
Is the literal comparison good enough, or would `base_type_eq` (or `type_eq`) be required? I don't know whether symbol types could be in place here. But then a redundant type cast isn't that much of an issue, admittedly.